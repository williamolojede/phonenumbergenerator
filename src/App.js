import React, { Component } from 'react';

import NumbersGeneratorForm from './components/NumbersGeneratorForm';
import GeneratedNumberStats from './components/GeneratedNumbersStats';
import GeneratedNumbersTable from './components/GeneratedNumbersTable';
import Sort from './components/Sort';

class App extends Component {
  state = {
    numbers: [],
    max: 0,
    min: 0,
    phoneNumbersGenerated: false
  }

  handleGeneratedNumbers = (numbers) => {
    this.setState({
      numbers,
      min: Math.min(...numbers),
      max: Math.max(...numbers),
      phoneNumbersGenerated: true
    })
  }

  handleSort = (dir) => () => {
    const { numbers } = this.state;

    numbers.sort((a, b) => {
      const sortDirs = {
        'asc': a - b,
        'dsc': b - a,
      }
      return sortDirs[dir];
    })
  
    this.setState({ numbers });
  }

  render() {
    const { numbers, max, min, phoneNumbersGenerated } = this.state;
    return (
      <div className="bg-grey-lighter min-h-screen">
        <header className="bg-purple-darker p-6 shadow-md">
          <div className="container mx-auto">
            <h3 className="text-white  text-center uppercase text-3xl">Phone number generator</h3>
          </div>
        </header>
        <main className="p-6 flex flex-col md:flex-row items-center md:items-start">
          <section className="container md:max-w-xs">
            <div className="bg-white shadow-md p-6 rounded-sm">
              
              <NumbersGeneratorForm handleGeneratedNumbers={this.handleGeneratedNumbers}/>
              {
                phoneNumbersGenerated && (
                  <React.Fragment>
                    <hr className="h-px my-6 bg-grey-light"/>

                    <GeneratedNumberStats count={numbers.length} min={min} max={max} />
    
                    <hr className="h-px my-6 bg-grey-light"/>

                    <Sort handleSort={this.handleSort}/>
                  </React.Fragment>
                )
              }
            </div>
          </section>
          <section className="container md:ml-6">
            <h3 className="py-6 md:pt-0 uppercase text-center">phone number list</h3>
            <GeneratedNumbersTable numbers={numbers}/>
          </section>
        </main>
      </div>
    );
  }
}

export default App;
