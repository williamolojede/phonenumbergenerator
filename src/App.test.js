import React from 'react'
import { mount } from 'enzyme';

import App from './App';

it('renders without crashing', () => {
  const wrapper = mount(<App />)
  expect(wrapper).toMatchSnapshot();
  const numbers = [
    100494045,
    100547806,
    101628526,
    102369825,
    105988240
  ]
  wrapper.instance().handleGeneratedNumbers(numbers)
  const { numbers: actualNumbers, max, min, phoneNumbersGenerated }  = wrapper.state()
  expect(numbers).toMatchObject(actualNumbers)
  expect(max).toBe(Math.max(...numbers))
  expect(min).toBe(Math.min(...numbers))
  expect(phoneNumbersGenerated).toBe(true)

});

describe('handleSort()', () => {
  const numbers = [
    100547806,
    102369825,
    100494045,
    105988240,
    101628526,
  ];

  it('should sort numbers in ascending order', () => {
    const wrapper = mount(<App />)
    const expectedSortedNumbers = [...numbers].sort((a, b) => a - b);
  
    wrapper.setState({ numbers: [...numbers] });
    wrapper.instance().handleSort('asc')();
  
    const { numbers: actualSortedNumbers }  = wrapper.state();
    expect(actualSortedNumbers).toMatchObject(expectedSortedNumbers);
  });

  it('should sort numbers in descending order', () => {
    const wrapper = mount(<App />)
    const expectedSortedNumbers = [...numbers].sort((a, b) => b - a);
  
    wrapper.setState({ numbers: [...numbers] });
    wrapper.instance().handleSort('dsc')();
  
    const { numbers: actualSortedNumbers }  = wrapper.state();
    expect(actualSortedNumbers).toMatchObject(expectedSortedNumbers);
  });
})

