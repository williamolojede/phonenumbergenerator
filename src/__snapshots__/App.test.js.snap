// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<App>
  <div
    className="bg-grey-lighter min-h-screen"
  >
    <header
      className="bg-purple-darker p-6 shadow-md"
    >
      <div
        className="container mx-auto"
      >
        <h3
          className="text-white  text-center uppercase text-3xl"
        >
          Phone number generator
        </h3>
      </div>
    </header>
    <main
      className="p-6 flex flex-col md:flex-row items-center md:items-start"
    >
      <section
        className="container md:max-w-xs"
      >
        <div
          className="bg-white shadow-md p-6 rounded-sm"
        >
          <NumbersGeneratorForm
            handleGeneratedNumbers={[Function]}
          >
            <form
              onSubmit={[Function]}
            >
              <p
                className="uppercase mb-6 text-center text-grey-dark"
              >
                Enter the number of phone numbers to be generated
              </p>
              <div
                className="flex justify-between"
              >
                <input
                  className="border-b-2 border-grey-darker text-5xl w-32 text-center"
                  max="500"
                  min="10"
                  required={true}
                  type="number"
                />
                <button
                  className="border-2 border-grey-darker p-4 w-1/2 text-xl font-bold uppercase text-grey-darker"
                  type="submit"
                >
                  Generate
                </button>
              </div>
            </form>
          </NumbersGeneratorForm>
        </div>
      </section>
      <section
        className="container md:ml-6"
      >
        <h3
          className="py-6 md:pt-0 uppercase text-center"
        >
          phone number list
        </h3>
        <GeneratedNumbersTable
          numbers={Array []}
        >
          <table
            className="w-full shadow-md"
          >
            <thead
              className="bg-grey-light"
            >
              <tr
                className="text-grey-dark"
              >
                <th
                  className="py-3"
                >
                  COUNT
                </th>
                <th
                  className="py-3"
                >
                  PHONE NUMBER
                </th>
              </tr>
            </thead>
            <tbody />
          </table>
        </GeneratedNumbersTable>
      </section>
    </main>
  </div>
</App>
`;
